tsdb/exemplar.go

```diff
@@ -0,0 +1,135 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// single circular buffer for all exemplars
+type CircularExemplarStorage struct {
+	lock      sync.RWMutex
+	index     map[string]int
+	exemplars []circularBufferEntry
+	nextIndex int
+	len       int
+}
+
+type circularBufferEntry struct {
+	exemplar     exemplar.Exemplar
+	seriesLabels labels.Labels // need to store labels so we can double check when querying
+	prev         int           // index of previous exemplar in circular for the same series
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int) *CircularExemplarStorage {
+	return &CircularExemplarStorage{
+		exemplars: make([]circularBufferEntry, len),
+		index:     make(map[string]int),
+		len:       len,
+	}
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+// TODO: separate wrapper struct for queries?
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(hints *storage.SelectHints, l labels.Labels) ([]exemplar.Exemplar, error) {
+	var (
+		ret []exemplar.Exemplar
+		e   exemplar.Exemplar
+		idx int
+		ok  bool
+		buf []byte
+	)
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	if idx, ok = ce.index[l.String()]; !ok {
+		return nil, nil
+	}
+	lastTs := ce.exemplars[idx].exemplar.Ts
+
+	for {
+		if idx == -1 || string(ce.exemplars[idx].seriesLabels.Bytes(buf)) != string(l.Bytes(buf)) {
+			break
+		}
+
+		e = ce.exemplars[idx].exemplar
+		// This line is needed to avoid an infinite loop, consider redesign of buffer entry struct.
+		if e.Ts > lastTs {
+			break
+		}
+
+		lastTs = ce.exemplars[idx].exemplar.Ts
+		// Prepend since this exemplar came before the last one we appeneded chronologically.
+		if e.Ts >= hints.Start && e.Ts <= hints.End {
+			ret = append([]exemplar.Exemplar{e}, ret...)
+		}
+		idx = ce.exemplars[idx].prev
+	}
+	return ret, nil
+}
+
+func (ce *CircularExemplarStorage) AddExemplar(l labels.Labels, t int64, e exemplar.Exemplar) error {
+	ce.lock.RLock()
+	idx, ok := ce.index[l.String()]
```

Comment:
You might want to call `l.String()` once and use it everywhere. Each one leads to a new string I believe.